docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md

```diff
@@ -0,0 +1,366 @@
+All files are concentrated under a single directory, whose name is hard coded. So actions such as "reset" are mapped to the deletion of all files under this directory, while iteration actions use file system APIs to traverse the directory.  
+Actual data is followed by a metadata structure. It holds the (currently unused) flags and data size, and its purpose is to verify streaming set was correctly finalized.     
+
+ 
+# Detailed design
+
+FileSystemStore fully implements the KVStore interface over a file system. As such, it uses the FileSystem class interface for file operations. 
+
+![FileSystemStore Class Hierarchy](./FileSystemStore_class_hierarchy.jpg)
+
+Functionality, as defined by KVStore, includes the following:
+- Initialization & reset
+- Core actions: get, set & remove
+- Incremental set actions
+- Iterator actions
+
+### Class header
+
+FileSystemStore has the following header:
+
+```C++
+class FileSystemStore : KVStore {
+
+public:
+    FileSystemStore(size_t max_keys, FileSystem *fs = 0);
+    virtual ~FileSystemStore();
+    	 
+    // Initialization and reset
+    virtual int init();
+    virtual int deinit();
+    virtual int reset();
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *pattern=NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);
+    virtual int iterator_close(iterator_t it);
+    
+private:
+    Mutex _mutex;
+    size_t *_max_keys;
+    size_t *_num_keys;
+    FileSystem *_fs;
+    bool _is_initialized;
+}
+```
+
+### Important data structures
+
+```C++
+// Key metadata
+typedef struct {
+    uint32_t magic;
+    uint16_t metadata_size;
+    uint16_t revision;
+    uint32_t ext_flags;
+    uint32_t data_size;
+} key_metadata_t;
+
+// incremental set handle
+typedef struct {
+    char *key;
+    uint32_t create_flags;
+} inc_set_handle_t;
+
+// iterator handle
+typedef struct {
+    void *dir_handle;
+    char *pattern;
+} key_iterator_handle_t;
+```
+
+
+### Initialization and reset
+
+**init function**
+
+Header:  
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
```

Comment:
Should _mutex also protect this if check?
If so, I think _mutex should be created at constructor or statically otherwise it wouldn't be able to protect init().

Reply:
Well, pseudo code doesn't show all code, but basic concepts. May put this initialization in the constructor as you suggested, or use an atomic counter as done in NVStore.

Reply:
One big value of pseudo code will be the security review. I think this should be the criteria to add or remove functionality from the pseudo code